src/main/osd/osd.c

```diff
@@ -1300,44 +1305,51 @@ void osdUpdate(timeUs_t currentTimeUs)
 
         osdSyncBlink();
 
-        uint8_t elementGroup;
-        uint8_t activeElements = osdGetActiveElementCount();
+        osdState = OSD_STATE_GROUP_ELEMENTS;
+
+        break;
+
+    case OSD_STATE_GROUP_ELEMENTS:
+        {
+            uint8_t elementGroup;
+            uint8_t activeElements = osdGetActiveElementCount();
 
-        // Reset groupings
-        for (elementGroup = 0; elementGroup < OSD_GROUP_COUNT; elementGroup++) {
-            if (osdElementGroupDurationUs[elementGroup] > (osdElementGroupTargetUs[elementGroup] + OSD_ELEMENT_RENDER_GROUP_MARGIN)) {
-                osdElementGroupDurationUs[elementGroup] = 0;
+            // Reset groupings
+            for (elementGroup = 0; elementGroup < OSD_GROUP_COUNT; elementGroup++) {
+                if (osdElementGroupDurationFracUs[elementGroup] > (OSD_ELEMENT_RENDER_TARGET << OSD_EXEC_TIME_SHIFT)) {
+                    osdElementGroupDurationFracUs[elementGroup] = 0;
+                }
+                osdElementGroupTargetFracUs[elementGroup] = 0;
             }
-            osdElementGroupTargetUs[elementGroup] = 0;
-        }
 
-        elementGroup = 0;
-
-        // Based on the current element rendering, group to execute in approx 40us
-        for (uint8_t curElement = 0; curElement < activeElements; curElement++) {
-            if ((osdElementGroupTargetUs[elementGroup] == 0) ||
-                ((osdElementGroupTargetUs[elementGroup] + osdElementDurationUs[curElement]) <= OSD_ELEMENT_RENDER_TARGET) ||
-                (elementGroup == (OSD_GROUP_COUNT - 1))) {
-                osdElementGroupTargetUs[elementGroup] += osdElementDurationUs[curElement];
-                // If group membership changes, reset the stats for the group
-                if (osdElementGroupMembership[curElement] != elementGroup) {
-                    osdElementGroupDurationUs[elementGroup] = 0;
+            elementGroup = 0;
+
+            // Based on the current element rendering, group to execute in approx 40us
+            for (uint8_t curElement = 0; curElement < activeElements; curElement++) {
+                if ((osdElementGroupTargetFracUs[elementGroup] == 0) ||
+                    (osdElementGroupTargetFracUs[elementGroup] + (osdElementDurationUs[curElement]) <= (OSD_ELEMENT_RENDER_TARGET << OSD_EXEC_TIME_SHIFT)) ||
+                    (elementGroup == (OSD_GROUP_COUNT - 1))) {
+                    osdElementGroupTargetFracUs[elementGroup] += osdElementDurationUs[curElement];
+                    // If group membership changes, reset the stats for the group
+                    if (osdElementGroupMembership[curElement] != elementGroup) {
+                        osdElementGroupDurationFracUs[elementGroup] = osdElementGroupTargetFracUs[elementGroup] + (OSD_ELEMENT_RENDER_GROUP_MARGIN << OSD_EXEC_TIME_SHIFT);
+                    }
+                    osdElementGroupMembership[curElement] = elementGroup;
```

Comment:
a quick word on naming.  Abbreviations seem to be overlooked these days, the idea is that you must be able to speak out loud names of things so that you can discuss them with other humans without sounding weird and so that you can be understood in plain english.

'cur' and 'frac' are recent examples that I've seen which don't follow the desired naming convention in the codebase which I've been continually trying to improve from the multiwii and baseflight origins.

in these to cases, it's better to have 'current' and 'fraction' instead of 'cur' and 'frac'.  It also helps non-english native contributors understand the codebase so they don't have to guess what the abbreviations mean.

also, a quick note about plurals,  arrays should be plural.  so that when you use them it makes sense.  e.g. get an `item` from the `items`.   i.e.  `uint8_t item = items[index];`

`osdElementGroupMembership[curElement] = elementGroup` should probably be `osdElementGroupMemberships[currrentElement] = elementGroup`

Reply:
I'm all for maintainability/clarity. Done.